state/cleanup.go

```diff
@@ -752,10 +785,64 @@ func (st *State) cleanupUnitsForDyingApplication(applicationname string, cleanup
 	return nil
 }
 
+func (st *State) cancelCleanupForceDestroyedUnit(appName string) error {
+	if appName == "" {
+		return nil
+	}
+	shouldCancel := func() (bool, error) {
+		app, err := st.Application(appName)
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		units, err := app.AllUnits()
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		return len(units) == 0, nil
+	}
+	if should, err := shouldCancel(); err != nil {
+		return errors.Trace(err)
+	} else if !should {
+		// We probably should not return an error to fail current main cleanup.
+		return nil
+	}
+	// Cancel unit cleanups.
+	ops, err := st.cancelCleanupOps(cleanupForceDestroyedUnit, appName)
+	if err != nil {
+		return errors.Annotatef(err, "can not get %q cancelCleanupOps for %q", cleanupForceDestroyedUnit, appName)
+	}
+	if len(ops) == 0 {
+		return nil
+	}
+	if err := st.db().RunTransaction(ops); err != nil {
+		return errors.Annotatef(err, "cannot cancel %q cleanup for %q", cleanupForceDestroyedUnit, appName)
+	}
+	return nil
+}
+
 // cleanupCharm is speculative: it can abort without error for many
 // reasons, because it's triggered somewhat over-enthusiastically for
 // simplicity's sake.
-func (st *State) cleanupCharm(charmURL string) error {
+func (st *State) cleanupCharm(charmURL string, cleanupArgs []bson.Raw) (err error) {
```

Comment:
The cleanupArgs would be set in `finalAppCharmRemoveOps` and I can't see that we're doing that in this PR?
In any case, do we need to tie those ops the charm cleanup? Can't we just add the `cancelCleanupOps` to the app `removeOps` directly? We already have a number of other remove ops added in there, eg `removeStatusOp`, `removeLeadershipSettingsOp`, `annotationRemoveOp` etc. That way we are removing the cleanup ops at the same time the final app artifacts are removed.

Reply:
yeah, we cancel the `cleanupForceDestroyedUnit` docs in app.removeOps now.